src/App/Fossa/Analyze.hs

```diff
@@ -408,15 +408,22 @@ analyze cfg = Diag.context "fossa-analyze" $ do
   let keywordSearchResultsFound = (maybe False (not . null . lernieResultsKeywordSearches) lernieResults)
   let outputResult = buildResult includeAll additionalSourceUnits filteredProjects' licenseSourceUnits
 
-  -- If we find nothing but keyword search, we exit with an error, but explain that the error may be ignorable.
-  -- We do not want to succeed, because nothing gets uploaded to the API for keyword searches, so `fossa test` will fail.
-  -- So the solution is to still fail, but give a hopefully useful explanation that the error can be ignored if all you were expecting is keyword search results.
-  case (keywordSearchResultsFound, checkForEmptyUpload includeAll projectScans filteredProjects' additionalSourceUnits licenseSourceUnits) of
-    (False, NoneDiscovered) -> Diag.fatal ErrNoProjectsDiscovered
-    (True, NoneDiscovered) -> Diag.fatal ErrOnlyKeywordSearchResultsFound
-    (False, FilteredAll) -> Diag.fatal ErrFilteredAllProjects
-    (True, FilteredAll) -> Diag.fatal ErrOnlyKeywordSearchResultsFound
-    (_, CountedScanUnits scanUnits) -> doUpload outputResult iatAssertion destination basedir jsonOutput revision scanUnits
+  scanUnits <-
+    case (keywordSearchResultsFound, checkForEmptyUpload includeAll projectScans filteredProjects' additionalSourceUnits licenseSourceUnits) of
+      (False, NoneDiscovered) -> do
+        Diag.warn ErrNoProjectsDiscovered
+        pure emptyScanUnits
+      (True, NoneDiscovered) -> do
```

Comment:
[Optional]:

```haskell
do   Diag.warn ErrOnlyKeywordSearchResultsFound
        pure emptyScanUnits
```
Is equivalent to:
```haskell
Diag.warn ErrOnlyKeywordSearchResultsFound $> emptyScanUnits
```
The type of `$>` is `$> :: f a -> b -> f b`. Use whatever seems more readable to you. But `do` notation is technically not necessary here because while diagnostics _are_ a monad this code isn't using anything its applicative interface doesn't already provide.

Reply:
Oh nice, I like this better actually